### DISCUSSION TITLE
Testing out a move from odb to arcadedb, Had issues with indexing, connecting to docker and remote session.

### DISCUSSION THREAD

**Discussion by thoemmo** [ArcadeData/arcadedb#638] [Q&A]
Hi ArcadeDB community,

Im considering a move from OrientDB to ArcadeDB and therefore im testing out the latest arcadedb.
I ran into some trouble please see below

1)
I didn't manage to import records with the import tool, im thinking maybe its because im on an old odb version 2.37
I decided to write some import code (good arcadedb practice). With the embedded server i started by creating two vertices V (with a property mid as a string) and Company where company extends V (like odb) I also managed to insert about 12000 records. The problem comes when I create a unique index on the mid property:

Code for index creation:

 Database db = server.getOrCreateDatabase("test_v2");
 db.getSchema().createTypeIndex(Schema.INDEX_TYPE.LSM_TREE, true, "V", new String[] { "mid" });

Exception:

com.arcadedb.index.IndexException: Error on creating index on type 'V', properties [mid]
...
Caused by: com.arcadedb.exception.TransactionException: Transaction error on commit
	at com.arcadedb.database.TransactionContext.commit1stPhase(TransactionContext.java:518)
	at com.arcadedb.database.TransactionContext.commit(TransactionContext.java:114)
	at com.arcadedb.database.EmbeddedDatabase.lambda$commit$2(EmbeddedDatabase.java:364)
	at com.arcadedb.database.EmbeddedDatabase.executeInReadLock(EmbeddedDatabase.java:1346)
	at com.arcadedb.database.EmbeddedDatabase.commit(EmbeddedDatabase.java:359)
	at com.arcadedb.database.EmbeddedDatabase.transaction(EmbeddedDatabase.java:998)
	at com.arcadedb.schema.EmbeddedSchema.lambda$createTypeIndex$4(EmbeddedSchema.java:484)
	at com.arcadedb.database.EmbeddedDatabase.executeInWriteLock(EmbeddedDatabase.java:1372)
	at com.arcadedb.database.EmbeddedDatabase.recordFileChanges(EmbeddedDatabase.java:1392)
	at com.arcadedb.schema.EmbeddedSchema.recordFileChanges(EmbeddedSchema.java:1370)
	at com.arcadedb.schema.EmbeddedSchema.createTypeIndex(EmbeddedSchema.java:483)
	... 3 more
Caused by: com.arcadedb.exception.TransactionException: Transaction buffer bigger than 2.00GB. Split the big transaction in smaller transactions. This transaction will be roll backed
	at com.arcadedb.engine.WALFile.writeTransactionToBuffer(WALFile.java:231)
	at com.arcadedb.engine.TransactionManager.createTransactionBuffer(TransactionManager.java:138)
	at com.arcadedb.database.TransactionContext.commit1stPhase(TransactionContext.java:507)
	
	
I dont understand how I can split index creation in smaller transactions. 
Currently in odb I have many more records and indices without problems so maybe there is an issue here

2)
I have some problems when I try to connect to the latest docker container

This line of code:

RemoteDatabase remoteDB = new RemoteDatabase("localhost", 2480, "test_v2", "root", "playwithdata");

Gives this error:

Error on executing remote command 'server' (httpErrorCode=405 httpErrorDescription=Method Not Allowed reason=null detail=null exception=null)

If I change the following in RemoteDatabase.java line 457 :

private void requestClusterConfiguration() {
    serverCommand("POST", "server", "SQL", null, null, false, false, new Callback() {
//    serverCommand("GET", "server", "SQL", null, null, false, false, new Callback() {
...

Then i can connect to the docker container, but I cant connect to ArcadeDBServer if I run it without docker. (so it seems to me the latest docker prefers POST and ArcadeDBServer prefers GET)

3)
If I do many inserts with RemoteDatabase my inserts will start to fail after a while

for (Vertex orientDBVertex : executeIterator) {

	String type = orientDBVertex.getProperty("@class");
        String json = ((OrientVertex) orientDBVertex).getRecord().toJSON("");
        
        try {

            arcadeDB.transaction(() -> {
                ResultSet res = arcadeDB.command("SQL", "insert into " + type + " content " + json);
            }, MAX_RETRIES);

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

Some of the records inserted will have lexical error because of json values containing either \", \u000a or \u000d. But once there is a lexical exception like:
Error on executing remote operation command (cause:com.arcadedb.query.sql.parser.TokenMgrException detail:Lexical error at line 1, column 647.  Encountered: " n" (10), after : " " u00d6 Vallgatan 57")

then the rest of the record insertion will start to fail. On the server there will be errors like this:

com.arcadedb.exception.DatabaseOperationException: Transaction not begun
at com.arcadedb.database.EmbeddedDatabase.checkTransactionIsActive(EmbeddedDatabase.java:1159)
at com.arcadedb.database.EmbeddedDatabase.lambda$rollback$3(EmbeddedDatabase.java:346)
at com.arcadedb.database.EmbeddedDatabase.executeInReadLock(EmbeddedDatabase.java:1299)
at com.arcadedb.database.EmbeddedDatabase.rollback(EmbeddedDatabase.java:344)
at com.arcadedb.server.ServerDatabase.rollback(ServerDatabase.java:129)
at com.arcadedb.server.http.handler.PostRollbackHandler.execute(PostRollbackHandler.java:34)
at com.arcadedb.server.http.handler.DatabaseAbstractHandler.execute(DatabaseAbstractHandler.java:82)
at com.arcadedb.server.http.handler.AbstractHandler.handleRequest(AbstractHandler.java:108)
at io.undertow.server.RoutingHandler.handleRequest(RoutingHandler.java:93)
at io.undertow.server.handlers.PathHandler.handleRequest(PathHandler.java:104)
at io.undertow.server.Connectors.executeRootHandler(Connectors.java:387)
at io.undertow.server.protocol.http.HttpReadListener.handleEventWithNoRunningRequest(HttpReadListener.java:256)
at io.undertow.server.protocol.http.HttpReadListener.handleEvent(HttpReadListener.java:136)
at io.undertow.server.protocol.http.HttpOpenListener.handleEvent(HttpOpenListener.java:162)
at io.undertow.server.protocol.http.HttpOpenListener.handleEvent(HttpOpenListener.java:100)
at io.undertow.server.protocol.http.HttpOpenListener.handleEvent(HttpOpenListener.java:57)
at org.xnio.ChannelListeners.invokeChannelListener(ChannelListeners.java:92)
at org.xnio.ChannelListeners$10.handleEvent(ChannelListeners.java:291)
at org.xnio.ChannelListeners$10.handleEvent(ChannelListeners.java:286)
at org.xnio.ChannelListeners.invokeChannelListener(ChannelListeners.java:92)
at org.xnio.nio.QueuedNioTcpServer2.acceptTask(QueuedNioTcpServer2.java:178)
at org.xnio.nio.WorkerThread.safeRun(WorkerThread.java:612)
at org.xnio.nio.WorkerThread.run(WorkerThread.java:479)

And on the client I will get 'com.arcadedb.exception.TransactionException: Transaction already begun'

By adding a 'sessionId = null' in RemoteDatabase.java line 150 (in the catch clause around rollback()) i can get what i think is expected behaviour, not sure that is actually how it shoud be fixed. The code below shows what i added:

  public void transaction(final Database.TransactionScope txBlock, int attempts) {
    if (txBlock == null)
      throw new IllegalArgumentException("Transaction block is null");

    ArcadeDBException lastException = null;

    if (attempts < 1)
      attempts = 1;

    for (int retry = 0; retry < attempts; ++retry) {
      try {
        begin();
        txBlock.execute();
        commit();

        return;
      } catch (NeedRetryException | DuplicatedKeyException e) {
        // RETRY
        lastException = e;
      } catch (Exception e) {
        try {
          rollback();
        } catch (Throwable t) {
          // IGNORE IT
          sessionId = null; //<------ Added this
        }
        throw e;
      }
    }

    throw lastException;
  }

4) When inserting json with insert queries are these characters not allowed: \", \u000a or \u000d? inside json values?

**Comment by thoemmo** [comment]
Please disregard  problem under 2) it is not an issue, wasn't using the latest after all

**Comment by thoemmo** [comment]
Got some more info for 1)
I actually have many more vertices extending V only there are no records added to those types (around 300 types in total). So 300 types with no records and one type called Data_Company with about 12000 records.
When I try to create the index  it fails with the exception shown above.

If I reduce the number of types to 10 then it succeeds and in the database folder a lot of files like this one gets created: XXX.umtidx  one for each bucket/type even if there are no records in those types. They are all 2.1MB. 

If I create the index over Data_Company instead of V I only get Data_Company_X_11776411128493.288.2097152.v1.umtidx files for Data_Company (8 in total) 

It seems like the number of types is important here even thought there are only records in one of the types!
(The field im trying to index is a uuid)

I have zero experience with db internals so I don't even know if this info is helpful, any way I hope to get some feedback!

**Comment by lvca** [comment]
About (3) the index creation does the whole job in one transaction, despite the number of records. This could cause an out-of-memory exception if the indexed records are a lot because the transaction keeps all the changes in RAM till the final `commit`.

There should be a new parameter/setting where the transaction is executed in chunks, namely multiple transactions. It should be managed also be the case when an error is raised and therefore the records indexed in the previous transactions should be removed.

About (4) do you have a test case to reproduce the issue?

**Reply by thoemmo** [reply]
I probably shouldnt have described all issues in one discussion. The memory issue is the one I call (1)
I probably didnt explain it right so i wrote some code to illustrate the problem:

TYPE_CNT is the number of Foo types
RECORD_CNT is the number of records inserted into the type Foo_1

if TYPE_CNT is 400 and RECORD_CNT is 1 I will run out of memory
if TYPE_CNT is 10 and RECORD_CNT is 1000000 it will index all records

```

    public static void main(String[] args) {

        DatabaseFactory arcade = new DatabaseFactory("/databases/Foo");
        if(arcade.exists()) {
            arcade.open().drop();
        }
        Database db = arcade.create();

        int TYPE_CNT = 400;
        int RECORD_CNT = 1;

        db.transaction(() -> {
            ResultSet res = db.command("SQL", "CREATE VERTEX TYPE V if not exists");
            VertexType vertexType = db.getSchema().getOrCreateVertexType("V");
            vertexType.getOrCreateProperty("id", Type.STRING);
        }, false, MAX_RETRIES);

        System.out.println();
        System.out.println("Types");
        System.out.println("-------------------------------------------------------------------");
        System.out.println("Done");
        db.transaction(() -> {
            for(int i = 0; i < TYPE_CNT; i++) {
                ResultSet res = db.command("SQL", "CREATE VERTEX TYPE Foo_" + i + " if not exists extends V");
            }
        }, false, MAX_RETRIES);

        System.out.println();
        System.out.println("Inserting");
        System.out.println("-------------------------------------------------------------------");
        db.transaction(() -> {
            for(int i = 0; i < RECORD_CNT; i++) {

                String json = "{\"id\":\"" + i + "\",\"content\":\"test test test test test test test test test test test test test\"}";
                ResultSet res = db.command("SQL", "insert into Foo_1 content " + json);
                System.out.println(i);
            }
        }, false, MAX_RETRIES);

        System.out.println();
        System.out.println("Idexing");
        System.out.println("-------------------------------------------------------------------");
        db.getSchema().createTypeIndex(Schema.INDEX_TYPE.LSM_TREE, true, "V", new String[] { "id" });
        System.out.println("Done");

        System.exit(0);
    }

```

With respect to (4)
If i run this test case under InsertStatementTest.java it will fail:

```
  @Test
  public void testSpecialCharactersInsert() {
    checkRightSyntax("insert into Foo content {\"test1\":\"Special character newline \\u000a\"}");
    checkRightSyntax("insert into Foo content {\"test1\":\"Special character carriage return \\u000d\"}");
    checkRightSyntax("insert into Foo content {\"test1\":\" double quotes \\\"\"}");
  } 
```

**Reply by lvca** [reply]
About (4) we fix an issue with unicode, you should try with the latest release.
About (3) it's a RAM problem: the transaction uses the RAM to keep the changes before committing. You should consider to split in multiple transactions (easier) or increase the available RAM
